File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlB1CancellationCleanupScenario.cs
// Verifies RL-B1 Cancellation Cleanup behavior.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-B1 verifies timeout/cancellation cleanup and a successful follow-up request.
internal static class RlB1CancellationCleanupScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        var firstMarker = $"rl-b1-first-{Guid.NewGuid():N}";
        await providerA.Post($"/admin/profile/hold/{firstMarker}").AsyncRaw();
        await providerB.Post($"/admin/profile/hold/{firstMarker}").AsyncRaw();

        using var firstCancellation = new CancellationTokenSource();
        using var evidenceTimeout = new CancellationTokenSource(TimeSpan.FromSeconds(15));
        var firstTask = consumer.Post("/profile/request")
            .Body(new ProfileReq("first", firstMarker))
            .Async<ProfileRes>(firstCancellation.Token)
            .AsTask();
        var waitA = providerA.Post("/evidence/wait")
            .Body(new EvidenceWaitReq([$"profile-start|rid=api-a|marker={firstMarker}"], []))
            .Async<string[]>(evidenceTimeout.Token)
            .AsTask();
        var waitB = providerB.Post("/evidence/wait")
            .Body(new EvidenceWaitReq([$"profile-start|rid=api-b|marker={firstMarker}"], []))
            .Async<string[]>(evidenceTimeout.Token)
            .AsTask();
        var completedEvidence = await Task.WhenAny(waitA, waitB);
        await completedEvidence;

        firstCancellation.Cancel();
        await AssertCanceledAsync(firstTask);

        var followUpMarker = $"rl-b1-follow-up-{Guid.NewGuid():N}";
        var followUp = (await consumer.Post("/profile/request")
            .Body(new ProfileReq("fast", followUpMarker))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(followUp.Value == "profile:fast", "RL-B1 follow-up request failed after timeout.");

        await providerA.Post($"/admin/profile/release/{firstMarker}").AsyncRaw();
        await providerB.Post($"/admin/profile/release/{firstMarker}").AsyncRaw();
        evidenceTimeout.Cancel();

        using (var lateReplyTimeout = new CancellationTokenSource(TimeSpan.FromSeconds(15)))
        {
            var lateWaitA = providerA.Post("/evidence/wait")
                .Body(new EvidenceWaitReq(["profile-request|", $"marker={firstMarker}"], []))
                .Async<string[]>(lateReplyTimeout.Token).AsTask();
            var lateWaitB = providerB.Post("/evidence/wait")
                .Body(new EvidenceWaitReq(["profile-request|", $"marker={firstMarker}"], []))
                .Async<string[]>(lateReplyTimeout.Token).AsTask();
            var completed = await Task.WhenAny(lateWaitA, lateWaitB);
            var evidence = (await completed).Body;
            ZlinkStreamAssert.Ensure(
                evidence.Any(line => line.Contains($"marker={firstMarker}", StringComparison.Ordinal)),
                "RL-B1 first request completion evidence missing after gate release.");
        }

        var later = (await consumer.Post("/profile/request")
            .Body(new ProfileReq("fast", $"rl-b1-later-{Guid.NewGuid():N}"))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(later.Value == "profile:fast", "RL-B1 later request failed after slow completion.");

        Console.WriteLine("scenario RL-B1 passed");
    }

    private static async Task AssertCanceledAsync(Task firstTask)
    {
        try
        {
            await firstTask;
        }
        catch (OperationCanceledException)
        {
            return;
        }

        throw new InvalidOperationException("RL-B1 first request was not canceled.");
    }
}

File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-b1-cancellation-cleanup-scenario.ts
// RL-B1: Client cancellation 뒤 후속 request를 처리한다 시나리오를 검증한다.
import type { ProfileRes, TimeoutRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { profileReq } from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';

export async function runRlB1(options: ClientOptions): Promise<void> {
  const slowMarker = `rl-b1-slow-${Date.now()}-${Math.random().toString(16).slice(2)}`;
  const timeout = await postJson<TimeoutRes>(
    options.consumerUrl,
    '/profile/request/timeout/100',
    profileRequestWithValue('slow', slowMarker)
  );
  ensure(timeout.status === 408 && timeout.timedOut, 'RL-B1 expected the slow request to time out.');

  const followUpMarker = `rl-b1-follow-up-${Date.now()}-${Math.random().toString(16).slice(2)}`;
  const followUp = await postJson<ProfileRes>(options.consumerUrl, '/profile/request', profileReq(followUpMarker));
  ensure(followUp.value === 'profile:fast', 'RL-B1 follow-up request failed after timeout.');

  const evidence = await Promise.any([
    postJson<string[]>(options.providerAUrl, '/evidence/wait', {
      contains: `profile-request|rid=api-a|value=slow|marker=${slowMarker}`,
      timeoutMilliseconds: 15000
    }),
    postJson<string[]>(options.providerBUrl, '/evidence/wait', {
      contains: `profile-request|rid=api-b|value=slow|marker=${slowMarker}`,
      timeoutMilliseconds: 15000
    })
  ]);
  ensure(
    evidence.some((line) => line.includes('profile-request|') && line.includes(`marker=${slowMarker}`)),
    'RL-B1 slow request completion evidence missing.'
  );

  const later = await postJson<ProfileRes>(
    options.consumerUrl,
    '/profile/request',
    profileReq(`rl-b1-later-${Date.now()}-${Math.random().toString(16).slice(2)}`)
  );
  ensure(later.value === 'profile:fast', 'RL-B1 later request failed after slow completion.');

  console.log('scenario RL-B1 passed');
}

function profileRequestWithValue(value: string, marker: string): { readonly value: string; readonly marker: string } {
  return { value, marker };
}
